src/analyzer/analyzerbeats.cpp

```diff
@@ -270,25 +274,7 @@ void AnalyzerBeats::storeResults(TrackPointer tio) {
         return;
     }
 
-    // If the user prefers to replace old beatgrids with newly generated ones or
-    // the old beatgrid has 0-bpm then we replace it.
-    bool zeroCurrentBpm = pCurrentBeats->getBpm() == 0.0;
-    if (m_bPreferencesReanalyzeOldBpm || zeroCurrentBpm) {
-        if (zeroCurrentBpm) {
-            qDebug() << "Replacing 0-BPM beatgrid with a" << pBeats->getBpm()
-                     << "beatgrid.";
-        }
-        tio->setBeats(pBeats);
-        return;
-    }
-
-    // If we got here then the user doesn't want to replace the beatgrid but
-    // since the first beat is zero we'll apply the offset we just detected.
-    double currentFirstBeat = pCurrentBeats->findNextBeat(0);
-    double newFirstBeat = pBeats->findNextBeat(0);
-    if (currentFirstBeat == 0.0 && newFirstBeat > 0) {
-        pCurrentBeats->translate(newFirstBeat);
-    }
+    tio->setBeats(pBeats);
```

Comment:
How about a single early exit instead of multiple paths?

```
if (pCurrentBeats && pCurrentBeats->getBpm() > 0.0 && tio->isBpmLocked()) {
    qDebug() << "Track was BPM-locked as we were analyzing it. Aborting analysis.";
    return;
}
```

Then how about encapsulating the condition in `Track::isBpmValidAndLocked()` (or `Track::isBeatGridValidAndLocked()`?) that requires only a single lock to retrieve a consistent value?

Reply:
I have considered an even more stream lined approach. Why not handle this inside setBeats? Than we can't forget to check the log flag everywhere.

Reply:
This makes especially sense after merging https://github.com/mixxxdj/mixxx/pull/3668 where beats are const and setBeats() is the only way to change the beats.